timemachine/cpp/src/kernels/k_nonbonded.cuh

```diff
@@ -430,4 +430,156 @@ void __global__ k_nonbonded_unified(
+            RealType w_j = params[w_param_idx_j];
+
+            RealType cj_x = coords[atom_j_idx * 3 + 0];
+            RealType cj_y = coords[atom_j_idx * 3 + 1];
+            RealType cj_z = coords[atom_j_idx * 3 + 2];
+
+            RealType delta_x = ci_x - cj_x;
+            RealType delta_y = ci_y - cj_y;
+            RealType delta_z = ci_z - cj_z;
+            RealType delta_w = w_i - w_j;
+
+            delta_x -= bx * nearbyint(delta_x * inv_bx);
+            delta_y -= by * nearbyint(delta_y * inv_by);
+            delta_z -= bz * nearbyint(delta_z * inv_bz);
+
+            RealType d2ij = delta_x * delta_x + delta_y * delta_y + delta_z * delta_z + delta_w * delta_w;
+
+            if (d2ij < cutoff_squared) {
+                RealType u;
+                RealType delta_prefactor;
+                RealType ebd;
+                RealType dij;
+                RealType inv_dij;
+                RealType inv_d2ij;
+                compute_electrostatics<RealType, true>(
+                    1.0, qi, qj, d2ij, beta, dij, inv_dij, inv_d2ij, ebd, delta_prefactor, u);
+
+                // lennard jones force
+                if (eps_i != 0 && eps_j != 0) {
+                    RealType sig_grad;
+                    RealType eps_grad;
+                    compute_lj<RealType, true>(
+                        1.0, eps_i, eps_j, sig_i, sig_j, inv_dij, inv_d2ij, u, delta_prefactor, sig_grad, eps_grad);
+                }
+                // Store the atom by atom energy
+                block_energy_buffer[threadIdx.x] = FLOAT_TO_FIXED_ENERGY<RealType>(u);
+            }
+        }
+        // Sync to ensure the shared buffers are populated
+        __syncthreads();
+
+        block_energy_reduce<THREADS>(block_energy_buffer, threadIdx.x);
+
+        if (threadIdx.x == 0) {
+            output_energies[blockIdx.y * gridDim.x + blockIdx.x] += block_energy_buffer[0];
+        }
+        // Sync the threads so threads don't move on and stomp on the block energy buffer
+        __syncthreads();
+
+        atom_j_idx += gridDim.x * blockDim.x;
+        // Always zero before the end of the loop, in case one thread no longer loops
+        block_energy_buffer[threadIdx.x] = 0;
+    }
+}
+
+template <typename RealType, int BLOCK_SIZE>
+void __global__ k_accumulate_atom_energies_to_per_mol_energies(
+    const int target_atoms,
+    const int target_mols,
+    const int *__restrict__ mol_idxs,               // [target_atoms]
+    const int *__restrict__ mol_offsets,            // [target_mols + 1]
+    const __int128 *__restrict__ per_atom_energies, // [target_atoms, BLOCK_SIZE]
+    __int128 *__restrict__ per_mol_energies) {
+    const int idx = blockIdx.x * blockDim.x + threadIdx.x;
+
+    if (idx >= target_mols) {
+        return;
+    }
+
+    __int128 energy_accumulator = 0;
+
+    const int mol_start = mol_offsets[idx];
+    const int mol_end = mol_offsets[idx + 1];
+
+    // TBD Parallelize
+    for (int i = mol_start; i < mol_end; i++) {
```

Comment:
This is likely very bad in the case of a protein

Reply:
Realized that this doesn't probably matter since we probably aren't interested in proteins as a target molecule.

Reply:
Practically, we'd only be interested in applying something like this to waters or ions.